Deliverable 2

AlanMalagon opened the pull request:
-Search bar is now enabled
-VideoDetail view was added
    *It shows an iframe that let the user play the video
    *The video info is showed below the video player
    *The can now see videos related at the right of the view

AlanMalagon committed:
Deliverable 1

It regards mini-challenges 1 and 2

AlanMalagon committed:
Deliverable 1

Pull request accepted for the first two mini-challenges of the Wizeline's React Certification 2021

AlanMalagon committed:
Deliverable 2

-Search bar is now enabled
-VideoDetail view was added
    *It shows an iframe that let the user play the video
    *The video info is showed below the video player
    *The can now see videos related at the right of the view

PacoMojica commented on src/components/App/App.component.jsx:
```diff
@@ -7,36 +8,28 @@ import LoginPage from '../../pages/Login';
 import SecretPage from '../../pages/Secret';
 import Private from '../Private';
-import Fortune from '../Fortune';
+// import Fortune from '../Fortune';
```
You can delete this, I don't think you'll need it

PacoMojica commented on src/components/App/App.component.jsx:
```diff
@@ -7,36 +8,28 @@ import LoginPage from '../../pages/Login';
-      body.removeEventListener('click', rotateBackground);
-    };
-  }, []);
+  const [searchValue, setSearchValue] = useState('doggos');
```
This looks good, in the next challenge you'll have an easy time setting the context

PacoMojica commented on src/components/Header/MenuButton/MenuButton.component.jsx:
```diff
@@ -0,0 +1,15 @@
+export const MenuButton = () => {
+  return <Container data-testid="menuButton">
+        <StyledButton whileTap={buttonPressedStyle}>
+            <StyledSvg height="32" viewBox="0 0 512 512" width="32" data-testid="menuIcon">
+                <path d="m464.883 64.267h-417.766c-25.98 0-47.117 21.136-47.117 47.149 0 25.98 21.137 47.117 47.117 47.117h417.766c25.98 0 47.117-21.137 47.117-47.117 0-26.013-21.137-47.149-47.117-47.149z"/>
+                <path d="m464.883 208.867h-417.766c-25.98 0-47.117 21.136-47.117 47.149 0 25.98 21.137 47.117 47.117 47.117h417.766c25.98 0 47.117-21.137 47.117-47.117 0-26.013-21.137-47.149-47.117-47.149z"/>
+                <path d="m464.883 353.467h-417.766c-25.98 0-47.117 21.137-47.117 47.149 0 25.98 21.137 47.117 47.117 47.117h417.766c25.98 0 47.117-21.137 47.117-47.117 0-26.012-21.137-47.149-47.117-47.149z"/>
+            </StyledSvg>
```
If you have more SVGs you can move them to their own folder, this improves the readability of the code

PacoMojica commented on src/components/Header/ProfileButton/ProfileButton.component.jsx:
```diff
@@ -0,0 +1,14 @@
+export const ProfileButton = () => {
+    return <Container data-testid="profileButton">
+       <StyledButton whileTap={buttonPressedStyle}>            
+            <StyledSvg x="0px" y="0px"width="32px" height="32px" viewBox="0 0 45.532 45.532" data-testid="profileIcon">
+                <g>
+                    <path d="M22.766,0.001C10.194,0.001,0,10.193,0,22.766s10.193,22.765,22.766,22.765c12.574,0,22.766-10.192,22.766-22.765 S35.34,0.001,22.766,0.001z M22.766,6.808c4.16,0,7.531,3.372,7.531,7.53c0,4.159-3.371,7.53-7.531,7.53 c-4.158,0-7.529-3.371-7.529-7.53C15.237,10.18,18.608,6.808,22.766,6.808z M22.761,39.579c-4.149,0-7.949-1.511-10.88-4.012 c-0.714-0.609-1.126-1.502-1.126-2.439c ...
+                </g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g><g></g>
+            </StyledSvg>
```
As I said, moving the SVGs to their own folder, or at least a file can improve the code readability a lot, it's just a minor thing

PacoMojica commented on src/components/Header/Searchbar/Searchbar.component.jsx:
```diff
@@ -0,0 +1,31 @@
+  const [inputValue, setInputValue] = useState("");
+  const history = useHistory();
+
+  const handleSubmit = (e) =>{
+    e.preventDefault();
+    setSearchValue(inputValue);
+    history.push("/");
+  }
+
+  const handleChange = (e) => {
+    setInputValue(e.target.value);
+  };
+
+  const handleIconClick = () =>{
+    setSearchValue(inputValue);
+    history.push("/");
+  }
```
I like that you define the handler here instead of _inline_ in the markup

PacoMojica commented on src/components/Layout/Layout.component.jsx:
```diff
@@ -3,7 +3,9 @@ import React from 'react';
 
 function Layout({ children }) {
-  return <main className="container">{children}</main>;
+  return <main className="container">
+      {children}
+    </main>;
```
```suggestion
  return (
    <main className="container">
      {children}
    </main>
  );
```

PacoMojica commented on src/components/VideoDetail/RelatedVideos/RelatedVideos.component.jsx:
```diff
@@ -0,0 +1,31 @@
+
+    return <Container>
+        {   
+            videos!==null && videos.items.length>0 &&
+            videos.items.map(video =>                
```
When you have multiple conditions you can create a variable with an expressive name to know what's happening, for example:

```jsx
const hasVideos = videos !== null && videos.items.length > 0;
```
and when used in the mark up it's almost like natural language

```jsx
return (
  <Container>
    {hasVideos && videos.items.map(...)}
  </Container>
);
```

PacoMojica commented on src/components/VideoDetail/VideoInfo/VideoInfo.component.jsx:
```diff
@@ -0,0 +1,27 @@
+    closed: {height: 50}
+}
+
+export const VideoInfo = ({video}) => {
+    const [isOpen, setIsOpen] = useState(false);
+    if(video){
+        const {title, channelTitle, publishedAt, description} = video.snippet;
+
+        return <div>
+            <Title>{title}</Title>
+            <ChannelAndDatePublished>{`${channelTitle} • ${parseDatetime(publishedAt)}`}</ChannelAndDatePublished>
+            <Separator></Separator>
+            <DescriptionContainer animate={isOpen ? "open":"closed"} variants={variants}>
+                <DescriptionText>{description}</DescriptionText>
+            </DescriptionContainer>
+            <ShowLessMore onClick={()=>setIsOpen(!isOpen)}>Show {!isOpen && 'more'} {isOpen && 'less'}</ShowLessMore>
+        </div>
+    }
+    return <></>
```
Here you have a prop called `video`, and you have an `if` statement to check if the prop in truthy, in this case the parent is the one that should check that prop, for example:

```jsx
function Parent() {
  const video = getFromSomewhere();

  return (
    <div>
      {video && <VideoInfo video={video} />}
    </div>
  );
}
```

There is no reason to return an empty fragment, if you use the suggestion you have to update the places where you use the component and check if the prop is truthy before trying to render the `VideoInfo`

```suggestion
export const VideoInfo = ({video}) => {
    const [isOpen, setIsOpen] = useState(false);
    const {title, channelTitle, publishedAt, description} = video.snippet;

    return (
      <div>
          <Title>{title}</Title>
          <ChannelAndDatePublished>{`${channelTitle} • ${parseDatetime(publishedAt)}`}</ChannelAndDatePublished>
          <Separator></Separator>
          <DescriptionContainer animate={isOpen ? "open":"closed"} variants={variants}>
              <DescriptionText>{description}</DescriptionText>
          </DescriptionContainer>
          <ShowLessMore onClick={()=>setIsOpen(!isOpen)}>Show {!isOpen && 'more'} {isOpen && 'less'}</ShowLessMore>
      </div>
    );
```

PacoMojica commented on test-coverage.txt:
```diff
@@ -0,0 +1,43 @@
+-------------------------------------|----------|----------|----------|----------|-------------------|
+File                                 |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
+-------------------------------------|----------|----------|----------|----------|-------------------|
+All files                            |    93.18 |       50 |    92.31 |    93.18 |                   |
```
Here you have a high percentage on everything except the branch column, you can try to solve this by creating more cases, most of the time this percentage is low because the events in some components are untested, you can look at the files generated by jest and search for the untested code

PacoMojica reviewed (commented):
Hi Alan, I liked you challenge, great work!

## Acceptance Criteria
- [x] Videos in the Home View are fetched from the YouTube API and displayed correctly.
- [x] Users can search for YouTube videos using the search field on the header.
- [x] A video can be played from the **Video Details View** after clicking on it.
- [x] The video information and related videos list are displayed correctly on the **Video Details View**.
- [x] When a user clicks on a related video the video player, information and related videos list are updated.

### Bonus Points
- [x] Custom Hooks for API calls are implemented and tested correctly.
- [x] Testing coverage is above 60%. (Please include a screenshot of the code coverage report in your PR description).

AlanMalagon committed:
Deliverable 2 test fixes

-Test cases that had problems with the previous commit are now fixed
-Some file thar are not used anymore were removed from the project

AlanMalagon committed:
Merge pull request #2 from AlanMalagon/alanmalagon

Deliverable 2 and Deliverable 2 fixes

AlanMalagon committed:
Deliverable 3

-The search term is stored and retrieved from the Global Context
-The appearance theme is stored in the Global Context and now working.
-User reducer handles the value of both search term and theme.
-Test cases not covered before were implemented and fixed to fit the new implementations

AlanMalagon committed:
Merge pull request #3 from AlanMalagon/deliverable-3

Deliverable 3

AlanMalagon committed:
Login mock api implementation

-Styles were added to login screen in order to make it reponsive and betterlooking
-A simple login using a mock of an API was implemented.
-The global state now manages the session data so AuthProvider was removed.

AlanMalagon committed:
Add and remove to favorites implemented

-The user can now fav a video by clicking the heart icon at the right of the title
-If the video is in the list, clicking again the button will remove from the list
-The icon changes depending on the existence of the video inside the list of favorite videos.

AlanMalagon committed:
Favorite videos view implemented

-Favorite videos view was implemented and is now accesible by click the profile icon

AlanMalagon committed:
Favorite video view via private route implemented

-Now when the user clicks on a favorite video is redirected to a private view
-The view uses the components from the original detail view it just changes the implementation of the related videos to show the videos in the favorite videos playlist instead.

AlanMalagon committed:
Side menu implemented

-The button menu now has tunctionality opens a side menu
-The items of the side menu redirect to the respectives views they name

AlanMalagon committed:
Menu state is now handled by the global state

-Some components has new calls to the dispatch function of the global state to close the menu everytime they are clicked

AlanMalagon committed:
Login component now shows as a modal

-The login component was modified to work as a modal that is shown over whatever component the user is in
-A the logout button was moved to be inside the favorite videos view, aka the profile of the user for now.

AlanMalagon committed:
Auth0 implemented to login into the app

-I know is a bad practice  to share credentials in this way but in order to make this easier to test they are 'wizeline@domain.com' and 'Rocks!123'

AlanMalagon committed:
Simpler styles were applied to the NotFound page

AlanMalagon committed:
Favorite icon imlemented in videos on home component

AlanMalagon committed:
Fixes to tests

AlanMalagon committed:
Merge branch 'deliverable-3' of https://github.com/AlanMalagon/react-certification-2020

AlanMalagon committed:
Merge branch 'deliverable-4' of https://github.com/AlanMalagon/react-certification-2020

AlanMalagon committed:
Merge branch 'deliverable-4' of https://github.com/AlanMalagon/react-certification-2020 into alanmalagon

AlanMalagon committed:
Quickfix to favorite videos when trying to add them from the home view

AlanMalagon committed:
Merge branch 'deliverable-4' of https://github.com/AlanMalagon/react-certification-2020 into alanmalagon

AlanMalagon committed:
Quick fix for Favorites Video View and Favorites Detail View

AlanMalagon committed:
Merge branch 'deliverable-4' of https://github.com/AlanMalagon/react-certification-2020 into alanmalagon
